src/core/thread/dua_manager.cpp

```diff
@@ -184,6 +253,508 @@ otError DuaManager::Store(void)
+
+    delay = config.mReregistrationDelay > 1 ? Random::NonCrypto::GetUint16InRange(1, config.mReregistrationDelay) : 1;
+
+    if (mDelay.mFields.mReregistrationDelay == 0 || mDelay.mFields.mReregistrationDelay > delay)
+    {
+        mDelay.mFields.mReregistrationDelay = delay;
+        ScheduleTimer();
+        otLogDebgDua("update reregdelay %d", mDelay.mFields.mReregistrationDelay);
+    }
+
+exit:
+    return;
+}
+
+void DuaManager::UpdateCheckDelay(uint8_t aDelay)
+{
+    if (mDelay.mFields.mCheckDelay == 0 || mDelay.mFields.mCheckDelay > aDelay)
+    {
+        mDelay.mFields.mCheckDelay = aDelay;
+
+        otLogDebgDua("update checkdelay %d", mDelay.mFields.mCheckDelay);
+        ScheduleTimer();
+    }
+}
+
+void DuaManager::HandleNotifierEvents(Events aEvents)
+{
+    Mle::MleRouter &mle = Get<Mle::MleRouter>();
+
+    if (aEvents.Contains(kEventThreadRoleChanged))
+    {
+        if (mle.HasRestored())
+        {
+            UpdateReregistrationDelay();
+        }
+#if OPENTHREAD_CONFIG_DUA_ENABLE && OPENTHREAD_FTD
+        else if (mle.IsRouter())
+        {
+            // Wait for link establishment with neighboring routers.
+            UpdateRegistrationDelay(kNewRouterRegistrationDelay);
+        }
+        else if (mle.IsExpectedToBecomeRouter())
+        {
+            // Will check again In case the device decides to stay REED when jitter timeout expires.
+            UpdateRegistrationDelay(mle.GetRouterSelectionJitterTimeout() + 1);
+        }
+#endif
+    }
+
+    if (aEvents.ContainsAny(kEventIp6AddressAdded))
+    {
+        mRegistrationTask.Post();
+    }
+}
+
+void DuaManager::HandleBackboneRouterPrimaryUpdate(BackboneRouter::Leader::State               aState,
+                                                   const BackboneRouter::BackboneRouterConfig &aConfig)
+{
+    OT_UNUSED_VARIABLE(aConfig);
+
+    if (aState == BackboneRouter::Leader::kStateAdded || aState == BackboneRouter::Leader::kStateToTriggerRereg)
+    {
+        UpdateReregistrationDelay();
+    }
+}
+
+void DuaManager::HandleTimer(void)
+{
+    bool attempt = false;
+
+    otLogDebgDua("regdelay %d, reregdelay %d, checkdelay %d", mDelay.mFields.mRegistrationDelay,
+                 mDelay.mFields.mReregistrationDelay, mDelay.mFields.mCheckDelay);
+
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+    if (mDuaState != kNotExist && TimerMilli::GetNow() > mLastRegistrationTime + Mle::kDuaDadPeriod)
+    {
+        mDomainUnicastAddress.mPreferred = true;
+    }
+
+    if (mDelay.mFields.mRegistrationDelay > 0)
+    {
+        mDelay.mFields.mRegistrationDelay--;
+
+        if (mDelay.mFields.mRegistrationDelay == 0)
+        {
+            attempt = true;
+        }
+    }
+#endif
+
+    if (mDelay.mFields.mCheckDelay > 0)
+    {
+        mDelay.mFields.mCheckDelay--;
+
+        if (mDelay.mFields.mCheckDelay == 0)
+        {
+            attempt = true;
+        }
+    }
```

Comment:
Looks fine as is, but this may be simpler
```suggestion
    if ((mDelay.mFields.mCheckDelay > 0) && (--mDelay.mFields.mCheckDelay == 0))
    {
        attempt = true;
   }
```